File: matlab/tools/other_eeg_processing/pre_whiten.m
function data = pre_whiten(data)
% Pre-whiten the input data using linear regression.
%
% Parameters:
% - data (numeric array): Matrix representing data in samples X channels.
%
% Returns:
% - data (numeric array): Pre-whitened data matrix.
%
% Example:
% data = randn(100, 5);  % Replace with your actual data
% pre_whitened_data = pre_whiten(data)

p = inputParser;
addRequired(p, 'data', @isnumeric);
parse(p, data);

data = p.Results.data;

for j = 1:size(data,2)
    vals = data(:,j);    
    
    if sum(~isnan(vals)) == 0
        continue
    end

    mdl = fitlm(vals(1:end-1),vals(2:end));
    E = mdl.predict(vals(1:end-1)) - vals(2:end);

    if length(E) < length(vals)
        E = [E;nan(length(vals)-length(E),1)];
    end
    data(:,j) = E;
end
end

File: python/CNTtools/tools/pre_whiten.py
from sklearn.linear_model import LinearRegression
import numpy as np
from beartype import beartype


@beartype
def pre_whiten(data: np.ndarray) -> np.ndarray:
    """Pre-whiten the input data using linear regression.

    Args:
        data (np.ndarray): Matrix representing data. Each column is a channel, and each row is a time point.

    Returns:
        np.ndarray: Pre-whitened data matrix.
    """
    for i in range(data.shape[1]):
        vals = data[:, i].reshape(-1, 1)
        if np.sum(~np.isnan(vals)) == 0:
            continue
        model = LinearRegression().fit(vals[:-1, :], vals[1:, :])
        E = model.predict(vals[:-1, :]) - vals[1:, :]
        if len(E) < len(vals):
            E = np.concatenate([E, np.nan * np.zeros([len(vals) - len(E), 1])])
        data[:, i] = E.reshape(-1)

    return data
